rfd/0143-external-k8s-joining.md

```diff
@@ -0,0 +1,280 @@
+- `curl http://localhost:8080/openid/v1/jwks`
+
+In the case that an operator rotates the certificate authority used by their
+Kubernetes Cluster to sign Service Account JWTS, the `jwks` field will also
+need to be updated.
+
+#### Configuring a pod for `kubernetes-jwks` joining
+
+The Pod will require a service account with a role granting it the ability to
+call the TokenRequest endpoint for itself. This allows the creation of a token
+with an audience of our choosing in order to include a nonce during the join
+to prevent token reuse.
+
+E.g for a service account named: "argocd":
+
+```yaml
+apiVersion: rbac.authorization.k8s.io/v1beta1
+kind: ClusterRole
+metadata:
+  name: argocd-token-request
+rules:
+- apiGroups: [""]
+  resources:
+  - "serviceaccounts/token"
+  verbs:
+  - "create"
+  # Only grant the ability to create tokens for itself. Missing this field leads
+  # to a dangerously powerful role.
+  resourceNames:
+    - argocd
+---
+apiVersion: rbac.authorization.k8s.io/v1beta1
+kind: RoleBinding
+metadata:
+  name: argocd-token-review
+roleRef:
+  apiGroup: rbac.authorization.k8s.io
+  kind: ClusterRole
+  name: argocd-token-review
+subjects:
+- kind: ServiceAccount
+  name: argocd
+  namespace: default
+```
+
+### Implementation
+
+<-- TODO: Specs for RegisterUsingKubernetesJWKS RPC -->
+
+The `kubernetes-jwks` flow will leverage a challenge and response flow, similar
+to that implemented for IAM joining. The flow will run as follows:
+
+1. The client wishing to join calls the `RegisterUsingKubernetesJWKS` RPC. It
+  submits the name of the token it wants to use.
+2. The Auth Server sends a response with the audience that should be set in the 
+  token. This audience is a concatenation of the cluster name, and 24 bytes of
+  cryptographically random data encoded in base64 URL encoding, e.g:
+  `example.teleport.sh/YWJjZGVmZ2hpamtsbW5vcHFyc3R1dnd4`
+3. The client uses its Kubernetes Service Account to call TokenRequest. It
+  requests a JWT for its own service account, with the audience specified by
+  the Auth Server and with a short time-to-live.
+4. The client sends a further request up the stream including this token.
+5. The server validates:
+  a. The JWT is signed by the keys present in the `jwks` field of the token 
+    resource.
+  b. The JWT is not expired, and at the time of issue, the JWTs TTL was short
+   (e.g within a similar TTL to the 30s TTL we set)
+  c. The JWTs audience claim matches the one the server challenged the client
+    with.
+  d. The JWT includes the `kubernetes.io` claim which binds it to a specified
+    namespace and pod.
+  e. The JWT subject claim is for a service account, and this service account 
+    matches one of those configured in the allow rules of the token resource.
+6. If the JWT passes validation, certificates are signed and returned to the 
+  client.
+
+## Security Considerations
+
+It's important to note that this join method will be less secure than the 
```

Comment:
Is it really less secure? The `kubernetes` join method ensures the token is invalidated as soon as the pod is stopped. However it is vulnerable to replay attacks (this is mitigated by the fact the communication is in-cluster, to do a MiTM attack, you need to already break in the Kube network). The proposed `kubernetes-jwks` method is not vulnerable to a MiTM because of the challenge.

Reply:
I guess the only scenario that comes to mind is one in which the Kubernetes SA JWT CA has been compromised but the attacker does not know the state of the cluster (and hence would have to "guess" at a running pods name). This seems pretty far-fetched to me, since if you had the ability to issue any service account, it would probably be reasonable to assume you could find a SA with permissions to list the currently running pods.

Reply:
6e07820a3a